File: eXoGadget/server/php/src/gadgets/JsLibraryFeatureFactory.php
<?php
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 * 
 */

class JsLibraryFeatureFactory extends GadgetFeatureFactory {
	private $JsLibraryFeature;

	public function __construct($gadgetLibraries, $containerLibraries)
	{
		// since we don't do strict type checking, this is one constructor for both a class, or a array of classes
		$this->JsLibraryFeature = new JsLibraryFeature($gadgetLibraries, $containerLibraries);
	}

	public function create()
	{
		return $this->JsLibraryFeature;
	}

	public function getLibraries($context)
	{
		return $context == 'GADGET' ? $this->JsLibraryFeature->gadgetLibraries : $this->JsLibraryFeature->containerLibraries;
	}
}

class JsLibraryFeature extends GadgetFeature {
	public $containerLibraries = array();
	public $gadgetLibraries = array();

	public function __construct($gadgetLibraries, $containerLibraries)
	{
		// we have a single constructor for both a single and multiple libraries, so handle it in code instead
		if ($gadgetLibraries != null && is_array($gadgetLibraries)) {
			$this->gadgetLibraries = array_merge($this->gadgetLibraries, $gadgetLibraries);
		} elseif ($gadgetLibraries != null && $gadgetLibraries instanceof JsLibrary) {
			$this->gadgetLibraries[] = $gadgetLibraries;
		}
		if ($containerLibraries != null && is_array($containerLibraries)) {
			$this->containerLibraries = array_merge($this->containerLibraries, $containerLibraries);
		} elseif ($containerLibraries != null && $containerLibraries instanceof JsLibrary) {
			$this->containerLibraries[] = $containerLibraries;
		}
	}

	public function prepare($gadget, $context, $params)
	{
		// do nothing
	}

	public function process($gadget, $context, $params)
	{
		$libraries = array();
		$libraries = $context->getRenderingContext() == 'GADGET' ? $this->gadgetLibraries : $this->containerLibraries;
		foreach ($libraries as $library) {
			$gadget->addJsLibrary($library);
		}
	}
}

File: gadgets/shindig/java/gadgets/src/main/java/org/apache/shindig/gadgets/JsLibraryFeatureFactory.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */
package org.apache.shindig.gadgets;

import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

/**
 * Enables basic JsLibrary loading without having to define new classes.
 *
 * Usage:
 * JsLibrary.register(...);
 * JsLibrary mylib = JsLibrary.registered("name", "name");
 * GadgetFeatureRegistry.register("name", null,
 *     new JsLibraryFeatureFactory(mylib));
 */
public class JsLibraryFeatureFactory implements GadgetFeatureFactory {
  private final JsLibraryFeature feature;

  public GadgetFeature create() {
    return feature;
  }

  /**
   * @param libraries The libraries to serve when this feature is used.
   */
  public JsLibraryFeatureFactory(
      Map<RenderingContext, Map<String, List<JsLibrary>>> libraries) {
    this.feature = new JsLibraryFeature(libraries);
  }
  protected JsLibraryFeatureFactory() {
    feature = null;
  }
}

class JsLibraryFeature extends GadgetFeature {
  final Map<RenderingContext, Map<String, List<JsLibrary>>> libraries;

  /**
   * Creates a JsLibraryFeature with multiple gadget and/or container libraries.
   * @param libraries
   */
  public JsLibraryFeature(
      Map<RenderingContext, Map<String, List<JsLibrary>>> libraries) {
    // TODO: technically we should copy this, but since currently the callers
    // always pass us something that won't be modified anyway, we're safe.
    // Copying this structure is painful.
    this.libraries = Collections.unmodifiableMap(libraries);
  }

  /**
   * {@inheritDoc}
   *
   * If context is null, all libraries will be returned rather than just
   * the libraries for the specified context.
   */
  @Override
  public List<JsLibrary> getJsLibraries(GadgetContext context) {
    List<JsLibrary> libs = null;

    if (context == null) {
      // for this special case we return all JS libraries in a single list.
      libs = new LinkedList<JsLibrary>();
      for (Map.Entry<RenderingContext, Map<String, List<JsLibrary>>> i :
          libraries.entrySet()) {
        for (Map.Entry<String, List<JsLibrary>> e : i.getValue().entrySet()) {
          libs.addAll(e.getValue());
        }
      }
    } else {
      Map<String, List<JsLibrary>> contextLibs
          = libraries.get(context.getRenderingContext());
      if (contextLibs != null) {
        libs = contextLibs.get(context.getContainer());
        if (libs == null) {
          // Try default.
          libs = contextLibs.get(ContainerConfig.DEFAULT_CONTAINER);
        }
      }
    }

    if (libs == null) {
      return Collections.emptyList();
    }
    return libs;
  }

  @Override
  public boolean isJsOnly() {
    return true;
  }
}
